src/main/java/duke/Parser.java

```diff
@@ -0,0 +1,186 @@
+package duke;
+
+import java.io.IOException;
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.time.format.DateTimeParseException;
+
+import duke.exceptions.IncorrectArgumentException;
+import duke.exceptions.InvalidCommandException;
+import duke.exceptions.ShutdownException;
+
+/** 
+ * Class which deals with making sense of the user command. 
+ * Deconstructs a line into a String[] of arguments, with the first item being the cmd and the rest the arg.
+ * Afterwards, performs checks for the validity of cmd and arg and throws exceptions if not valid.
+ * If no exceptions are thrown, it calls the respective classes to execute the command. 
+ */
+class Parser {
+    Storage storage;
+    Ui ui;
+    TaskList tasks;
+    
+    /**
+     * Public constructor which creates a Parser. As it passes valid commands to be executed by the respective
+     * classes, a reference to the corresponding classes is passed in during construction.
+     * @param storage a reference to the underlying Storage object.
+     * @param ui a reference to the underlying Ui object.
+     * @param tasks a reference to the TaskList that the current application uses to hold Task entities.
+     */
+    public Parser(Storage storage, Ui ui, TaskList tasks) {
+        this.storage = storage;
+        this.ui = ui;
+        this.tasks = tasks;
+    }
+
+    /**
+     * Parses the entire comment line by verifying the command and arguments, and then passes on the command 
+     * to the corresponding methods.
+     * @param line the entire command line retrieved from Scanner.nextLine()
+     * @return A string describing duke's response.
+     * @throws IncorrectArgumentException when the wrong number or type of arguments are supplied to the command.
+     * @throws InvalidCommandException when a command is entered that does not exist.
+     * @throws IOException when the Storage object fails in saving the TaskList to file.
+     */
+    public String[] parse(String line) throws IncorrectArgumentException, InvalidCommandException, IOException, ShutdownException {
+        String[] cmd = line.split(" ", 2);
+        String[] response;
+        try {
+            switch (cmd[0].toLowerCase()) {
+            case "bye":
+                throw new ShutdownException(bye());
```

Comment:
Command 'bye' should not be throwing an exception, since it is a valid 'command'! Think of another way to exit the program.

<details>
  <summary>Hint</summary>
  Just check the input string in Duke class
</details>